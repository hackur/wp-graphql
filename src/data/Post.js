import { toGlobalId } from 'graphql-relay';
import Dataloader from 'dataloader';
import fetchData from 'data/utils';

// Dataloader expects IDs that can be read by the REST API

const path = process.env.WP_POSTS_ENDPOINT || 'graphql/v1/posts';
const postLoader = new Dataloader(ids =>
  fetchData(path, {
    qs: { include: ids, orderby: 'include', per_page: 100 },
  }).then(({ data: { body } }) => body)
);
const slugLoader = new Dataloader(slugs =>
  fetchData(path, { qs: { slug: slugs, orderby: 'slug', per_page: 100 } }).then(
    ({ data: { body } }) => body
  )
);

class Post {
  getID() {
    return toGlobalId(this.constructor.name, this.id);
  }

  static getEndpoint() {
    return path;
  }

  static async load(id) {
    const data = await postLoader.load(id);
    return data ? Object.assign(new Post(), data) : null;
  }

  static async loadBySlug(slug) {
    const data = await slugLoader.load(slug);
    return data ? Object.assign(new Post(), data) : null;
  }

  static async collection(args = {}) {
    const { data: { body, headers } } = await fetchData(path, { qs: args });
    return {
      total: headers['x-wp-total'],
      items: body.map(item => Object.assign(new Post(), item)),
    };
  }
}

export default Post;
