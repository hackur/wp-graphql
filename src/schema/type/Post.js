import { GraphQLObjectType, GraphQLList } from 'graphql';
import PostInterface from 'interface/Post';
import CategoryType from 'type/Category';
import TagType from 'type/Tag';
import commentConnection from 'connection/Comment';
import { globalIdField, slug, guid, link } from 'field/identifier';
import { title, content, excerpt } from 'field/content';
import { date, modified } from 'field/date';
import { commentStatus, pingStatus } from 'field/status';
import { type, template, format, sticky } from 'field/post';
import { featuredMedia } from 'field/media';
import metaField from 'field/meta';
import author from 'field/author';
import Category from 'data/Category';
import Tag from 'data/Tag';

const PostType = new GraphQLObjectType({
  name: 'Post',
  description: 'A read-only post object.',
  interfaces: [PostInterface],
  isTypeOf(post) {
    return post.type === 'post';
  },
  fields: () => ({
    id: globalIdField(),
    ...date,
    ...guid,
    ...modified,
    ...slug,
    ...type,
    ...link,
    ...title,
    ...content,
    ...excerpt,
    ...commentStatus,
    ...pingStatus,
    ...template,
    ...format,
    ...author,
    featuredMedia: featuredMedia(),
    meta: metaField(),
    // extra post fields
    ...sticky,
    categories: {
      type: new GraphQLList(CategoryType),
      description: 'The terms assigned to the object in the category taxonomy.',
      resolve: ({ categories }) => {
        if (categories.length) {
          return Category.loadMany(categories);
        }
        return null;
      },
    },
    tags: {
      type: new GraphQLList(TagType),
      description: 'The terms assigned to the object in the post_tag taxonomy.',
      resolve: ({ tags }) => {
        if (tags.length) {
          return Tag.loadMany(tags);
        }
        return null;
      },
    },
    comments: commentConnection,
  }),
});

export default PostType;
