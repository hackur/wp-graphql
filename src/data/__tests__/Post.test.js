import Post from 'data/Post';

jest.mock('../utils', () =>
  jest.fn(() => ({
    data: {
      body: [{ id: 1 }, { id: 2 }, { id: 3 }, { id: 4 }, { id: 5 }],
      headers: {
        'x-wp-total': 5,
      },
    },
  }))
);

describe('Test Post data access', () => {
  test('Get endpoint', () => {
    expect(Post.getEndpoint()).toMatchSnapshot();
  });

  test('Load a post', async () => {
    const post = await Post.load(13);
    expect(post.getID()).toMatchSnapshot();
    expect(post).toMatchSnapshot();
  });

  test('Load a post by slug', async () => {
    const post = await Post.loadBySlug('whatever');
    expect(post).toMatchSnapshot();
  });

  test('Load Post collection', async () => {
    const collection = await Post.collection();
    expect(collection).toMatchSnapshot();
  });
});
