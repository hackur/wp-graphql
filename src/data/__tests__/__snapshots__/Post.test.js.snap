// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Post data access Get endpoint 1`] = `"graphql/v1/posts"`;

exports[`Test Post data access Load Post collection 1`] = `
Object {
  "items": Array [
    Post {
      "id": 1,
    },
    Post {
      "id": 2,
    },
    Post {
      "id": 3,
    },
    Post {
      "id": 4,
    },
    Post {
      "id": 5,
    },
  ],
  "total": 5,
}
`;

exports[`Test Post data access Load a post 1`] = `"UG9zdDoxMw=="`;

exports[`Test Post data access Load a post 2`] = `
Post {
  "id": 13,
}
`;

exports[`Test Post data access Load a post by slug 1`] = `
Post {
  "slug": "whatever",
}
`;
